NadekoBot.Core/Modules/Utility/AutoComplete/QuoteAutoCompletehandler.cs

```diff
@@ -0,0 +1,46 @@
+using Discord;
+using Discord.Interactions;
+using NadekoBot.Common;
+using NadekoBot.Common.Attributes;
+using NadekoBot.Common.Replacements;
+using NadekoBot.Core.Services;
+using NadekoBot.Core.Services.Database.Models;
+using NadekoBot.Extensions;
+using System;
+using System.Threading.Tasks;
+using System.Linq;
+using System.Collections.Generic;
+using Serilog;
+namespace NadekoBot.Modules.Utility
+{
+    public class QuoteAutoCompleteHandler : AutocompleteHandler
+    {
+        IEnumerable<string> keywords = null;
+
+        public override Task<AutocompletionResult> GenerateSuggestionsAsync(IInteractionContext ctx, IAutocompleteInteraction autocompleteInteraction, IParameterInfo parameter, IServiceProvider services){
+            try
+            {
+                var value = autocompleteInteraction.Data.Current.Value as string;
+                value = value.ToUpperInvariant();
+                // No suggestions for strings less than 3 characters
+                if ((string.IsNullOrEmpty(value)) || (value.Length < 3)){
+                    return Task.FromResult(AutocompletionResult.FromSuccess());
+                }
+                if ((keywords == null) || (value.Length == 3)){
+                    // query for suggestions
+                    DbService db = (DbService)services.GetService(typeof(DbService));
+                    using (var uow = db.GetDbContext()){
+                        keywords = uow.Quotes.SearchDistinctQuoteKeywordAsync(ctx.Guild.Id,value);
+                    }
+                }
+                
+                var matches = keywords.Where(x=>x.StartsWith(value)).Take(25).ToList();
```

Comment:
The limit should be done at the sql level. That way it will have more consistent performance and not scale exactly linearly with the number of quotes

Reply:
No, it should be done here the fourth character has no guarantee to be in the top 25 quotes

Reply:
You can do the string filtering at the sql level too

Reply:
I already do the string filtering at the sql level.

Reply:
I'm missing something about how autocomplete works in discord. I thought it worked like

1. user types some fragment
2. quote keywords with that fragment as the prefix are shown

So the sql query is going to look like 
```sql
SELECT DISTINCT ... FROM ... WHERE keyword LIKE 'fooba%' LIMIT 25
```

It looks like all of the logic can be done in the SQL. Is there some other ranking or transformation being done to the suggestions that can't be done in SQL?

Reply:
No it's expensive to do SQL queries everytime a user types a character. I do a single SQL query when there is 3 character typed and get list of distinct keywords, then return top 25 from that list. If user types 4th character I use results from previous query and do further filtering.

Reply:
Do you have a feel for what the difference is? Is it 1s vs 0.1s, or is it 5s vs 0.5s?

Reply:
I have no clue, like before it is impossible for me to test this